cql3/expr/expression.cc

```diff
@@ -247,41 +252,61 @@ bool_or_null equal(const expression& lhs, const managed_bytes_opt& rhs_bytes, co
     return type_of(lhs)->equal(managed_bytes_view(lhs_bytes), managed_bytes_view(*rhs_bytes));
 }
 
-static std::optional<std::pair<managed_bytes, managed_bytes>> evaluate_binop_sides(const expression& lhs,
+static std::pair<managed_bytes_opt, managed_bytes_opt> evaluate_binop_sides(const expression& lhs,
                                                                                    const expression& rhs,
                                                                                    const oper_t op,
                                                                                    const evaluation_inputs& inputs) {
     raw_value lhs_value = evaluate(lhs, inputs);
     raw_value rhs_value = evaluate(rhs, inputs);
 
-    if (lhs_value.is_null() || rhs_value.is_null()) {
-        return std::nullopt;
-    }
-    managed_bytes lhs_bytes = std::move(lhs_value).to_managed_bytes();
-    managed_bytes rhs_bytes = std::move(rhs_value).to_managed_bytes();
+    managed_bytes_opt lhs_bytes = std::move(lhs_value).to_managed_bytes_opt();
+    managed_bytes_opt rhs_bytes = std::move(rhs_value).to_managed_bytes_opt();
     return std::pair(std::move(lhs_bytes), std::move(rhs_bytes));
 }
 
 bool_or_null equal(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
-    std::optional<std::pair<managed_bytes, managed_bytes>> sides_bytes =
+    std::pair<managed_bytes_opt, managed_bytes_opt> sides_bytes =
         evaluate_binop_sides(lhs, rhs, oper_t::EQ, inputs);
-    if (!sides_bytes.has_value()) {
+    if (!sides_bytes.first || !sides_bytes.second) {
         return bool_or_null::null();
     }
-    auto [lhs_bytes, rhs_bytes] = std::move(*sides_bytes);
+    auto [lhs_bytes, rhs_bytes] = std::move(sides_bytes);
 
-    return type_of(lhs)->equal(managed_bytes_view(lhs_bytes), managed_bytes_view(rhs_bytes));
+    return type_of(lhs)->equal(managed_bytes_view(*lhs_bytes), managed_bytes_view(*rhs_bytes));
 }
 
 bool_or_null not_equal(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
-    std::optional<std::pair<managed_bytes, managed_bytes>> sides_bytes =
+    std::pair<managed_bytes_opt, managed_bytes_opt> sides_bytes =
         evaluate_binop_sides(lhs, rhs, oper_t::NEQ, inputs);
-    if (!sides_bytes.has_value()) {
+    if (!sides_bytes.first || !sides_bytes.second) {
         return bool_or_null::null();
     }
-    auto [lhs_bytes, rhs_bytes] = std::move(*sides_bytes);
+    auto [lhs_bytes, rhs_bytes] = std::move(sides_bytes);
+
+    return !type_of(lhs)->equal(managed_bytes_view(*lhs_bytes), managed_bytes_view(*rhs_bytes));
+}
+
+static bool equal_special_null_impl(const expression& lhs, const expression& rhs, const evaluation_inputs& inputs) {
+   std::pair<managed_bytes_opt, managed_bytes_opt> sides_bytes =
```

Comment:
I would prefer to have something like
```
enum class null_handling_style {
   standard,
   special
};
compare(managed_bytes_opt, managed_bytes_opt, null_handling_style) -> optional<bool_or_null>;
```
and not copy-paste `equal`, `not_equal`, `is_one_of_special_null`

Reply:
I'm thinking we should move the null handling style into binary_operator, instead of adding new enums. It will make not changing </<=/>/>= meaning easier. Next revision will have it so.